src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -874,19 +872,23 @@ static ClusterState selectNextStateToProcess(Queue<ProcessClusterState> processN
         return stateToProcess.clusterState;
     }
 
-    /**
-     * In the case we follow an elected master the new cluster state needs to have the same elected master and
-     * the new cluster state version needs to be equal or higher than our cluster state version. If either conditions
-     * are true then the cluster state is dated and we should ignore it.
-     */
-    static boolean shouldIgnoreNewClusterState(ESLogger logger, ClusterState currentState, ClusterState newClusterState) {
+    static boolean checkWrongMaster(ESLogger logger, ClusterState currentState, ClusterState newClusterState) {
         if (currentState.nodes().masterNodeId() == null) {
             return false;
         }
         if (!currentState.nodes().masterNodeId().equals(newClusterState.nodes().masterNodeId())) {
-            logger.warn("received a cluster state from a different master then the current one, ignoring (received {}, current {})", newClusterState.nodes().masterNode(), currentState.nodes().masterNode());
+            logger.warn("received a cluster state from a different master then the current one, rejecting (received {}, current {})", newClusterState.nodes().masterNode(), currentState.nodes().masterNode());
             return true;
-        } else if (newClusterState.version() < currentState.version()) {
+        } else {
+            return false;
+        }
+    }
+
+    static boolean checkStaleVersion(ESLogger logger, ClusterState currentState, ClusterState newClusterState) {
+        if (currentState.nodes().masterNodeId() == null) {
```

Comment:
now that we split it into two methods, we need to re-introduce the master checks , like the comment says:
> received a cluster state from a different master then the current one, rejecting